src/java/org/apache/cassandra/cql3/restrictions/ClusteringElements.java

```diff
@@ -0,0 +1,449 @@
+     */
+    public ClusteringElements extend(ClusteringElements suffix)
+    {
+        // We cannot extend a Top or Bottom has those are only use for ranges and ranges endpoint should not be extended
+        if (this instanceof Top || this instanceof Bottom)
+            throw new UnsupportedOperationException("Range endpoints cannot be extended");
+
+        checkSuffix(suffix);
+
+        ImmutableList<? extends ColumnSpecification> newColumns = concat(columns, suffix.columns);
+        ImmutableList<ByteBuffer> newValues = concat(values, suffix.values);
+
+        return suffix instanceof Top ? new Top(newColumns, newValues)
+                                     : suffix instanceof Bottom ? new Bottom(newColumns, newValues)
+                                                                : new ClusteringElements(newColumns, newValues);
+    }
+
+    private void checkSuffix(ClusteringElements suffix)
+    {
+        // If the columns are not ColumnMetadata instances we are dealing a Token representation which cannot be extended
+        if (!(columns.get(0) instanceof ColumnMetadata))
+            throw new UnsupportedOperationException("Non partition key or clustering columns cannot be extended");
+
+        if (!suffix.isEmpty()) // suffix can be empty if equal to (top) or (bottom)
+        {
+            ColumnMetadata lastPrefixElement = ((ColumnMetadata) last(this.columns));
+            ColumnMetadata firstSuffixElement = ((ColumnMetadata) suffix.columns.get(0));
+            if (firstSuffixElement.kind != lastPrefixElement.kind)
+                throw new UnsupportedOperationException("Cannot extends elements with elements of a different kind");
+            if (firstSuffixElement.position() != lastPrefixElement.position() + 1)
+                throw new UnsupportedOperationException("Cannot extends elements with not consecutive elements");
+        }
+    }
+
+    private static <T> ImmutableList<T> concat(ImmutableList<? extends T> prefix, ImmutableList<? extends T> suffix)
+    {
+        return ImmutableList.<T>builderWithExpectedSize(prefix.size() + suffix.size())
+                            .addAll(prefix)
+                            .addAll(suffix)
+                            .build();
+    }
+
+    /**
+     * Returns a {@code RangeSet} representing all the possible {@code ClusteringElements} values.
+     * @return a {@code RangeSet} representing all the possible {@code ClusteringElements} values.
+     */
+    public static RangeSet<ClusteringElements> all()
+    {
+        TreeRangeSet<ClusteringElements> rangeSet = TreeRangeSet.create();
+        rangeSet.add(ALL);
+        return rangeSet;
+    }
+
+    /**
+     * Returns a {@code RangeSet} that contains all values less than or equal to endpoint.
+     * @return a {@code RangeSet} that contains all values less than or equal to endpoint.
+     */
+    public static RangeSet<ClusteringElements> atMost(ClusteringElements endpoint)
```

Comment:
Why not name these 'lessThanOrEquals'?

Reply:
To keep the same naming as Guava `Range` and avoid adding another one.